system/security/shared/seclib.hpp

```diff
@@ -337,53 +409,110 @@ enum secManagerType : int
-    virtual bool addResources(ISecUser & user, ISecResourceList * resources) = 0;
-    virtual bool addResourcesEx(SecResourceType rtype, ISecUser & user, ISecResourceList * resources, SecPermissionType ptype, const char * basedn) = 0;
-    virtual bool addResourceEx(SecResourceType rtype, ISecUser & user, const char * resourcename, SecPermissionType ptype, const char * basedn) = 0;
-    virtual bool getResources(SecResourceType rtype, const char * basedn, IResourceArray & resources) = 0;
-    virtual bool updateResources(ISecUser & user, ISecResourceList * resources) = 0;
-    virtual bool updateSettings(ISecUser & user, ISecPropertyList * resources, IEspSecureContext* secureContext) = 0;
-    virtual bool addUser(ISecUser & user) = 0;
-    virtual ISecUser * findUser(const char * username) = 0;
-    virtual ISecUser * lookupUser(unsigned uid) = 0;
-    virtual ISecUserIterator * getAllUsers() = 0;
-    virtual void getAllGroups(StringArray & groups, StringArray & managedBy, StringArray & descriptions ) = 0;
-    virtual bool updateUserPassword(ISecUser & user, const char * newPassword, const char* currPassword = 0) = 0;
-    virtual bool initUser(ISecUser & user) = 0;
-    virtual void setExtraParam(const char * name, const char * value) = 0;
-    virtual IAuthMap * createAuthMap(IPropertyTree * authconfig) = 0;
-    virtual IAuthMap * createFeatureMap(IPropertyTree * authconfig) = 0;
-    virtual IAuthMap * createSettingMap(IPropertyTree * authconfig) = 0;
-    virtual void deleteResource(SecResourceType rtype, const char * name, const char * basedn) = 0;
-    virtual void renameResource(SecResourceType rtype, const char * oldname, const char * newname, const char * basedn) = 0;
-    virtual void copyResource(SecResourceType rtype, const char * oldname, const char * newname, const char * basedn) = 0;
-    virtual void cacheSwitch(SecResourceType rtype, bool on) = 0;
-    virtual bool authTypeRequired(SecResourceType rtype) = 0;
-    virtual SecAccessFlags authorizeWorkunitScope(ISecUser & user, const char * filescope) = 0;
-    virtual bool authorizeWorkunitScope(ISecUser & user, ISecResourceList * resources) = 0;
-    virtual const char * getDescription() = 0;
-    virtual unsigned getPasswordExpirationWarningDays() = 0;
-    virtual bool createUserScopes() = 0;
-    virtual aindex_t getManagedScopeTree(SecResourceType rtype, const char * basedn, IArrayOf<ISecResource>& scopes) = 0;
-    virtual SecAccessFlags queryDefaultPermission(ISecUser& user) = 0;
-    virtual bool clearPermissionsCache(ISecUser & user) = 0;
-    virtual bool authenticateUser(ISecUser & user, bool * superUser) = 0;
-    virtual secManagerType querySecMgrType() = 0;
-    virtual const char* querySecMgrTypeName() = 0;
-    virtual bool logoutUser(ISecUser & user) = 0;
+    virtual uint64_t queryFeatures(SecurityFeatureSupportLevel level, IEspSecureContext* secureContext = nullptr) const = 0;
```

Comment:
Minor, but could you add comment that this returns a mask of SUF features?  Developers should not have to look at the implementation to figure that out

Reply:
In other code that is on hold at the moment, I included using statements to define singular and plural types (in this case, feature versus features). Then the signatures were very explicit in denoting whether a singular value or multiple values were expected as input or returned. That approach was not as well received as I would have hoped, but I think it would address your concern by making the code somewhat self documenting. I can add a comment too, but I think returning a type name other than uint64_t would be even better.